From 046e592cc3993b3a7f9092f6b14bb6bbd8285f4c Mon Sep 17 00:00:00 2001
From: Maxim Ostapenko <maxim.o@samsung.com>
Date: Wed, 17 Nov 2021 17:39:22 +0900
Subject: [PATCH 031/135] [caffe2/python] Support
 'workspace.AxdimmStarter(num_cpu_inferenceEngines)'

Support AXDIMM starter initialization in Caffe2, this allows us to
get rid of python bindings in AXDIMM library.

Signed-off-by: Maxim Ostapenko <maxim.o@samsung.com>
---
 caffe2/python/pybind_state_axdimm.cc | 9 +++++++++
 caffe2/python/workspace.py           | 4 ++++
 2 files changed, 13 insertions(+)

diff --git a/caffe2/python/pybind_state_axdimm.cc b/caffe2/python/pybind_state_axdimm.cc
index 555b9c79..235d99bb 100644
--- a/caffe2/python/pybind_state_axdimm.cc
+++ b/caffe2/python/pybind_state_axdimm.cc
@@ -14,6 +14,8 @@
 #include "caffe2/operators/operator_fallback_axdimm.h"
 #include "caffe2/python/pybind_state_registry.h"
 
+#include "api/axdimm.h"
+
 namespace caffe2 {
 namespace python {
 
@@ -29,10 +31,17 @@ REGISTER_BLOB_FEEDER(AXDIMM, TensorFeeder<AXDIMMContext>);
 
 namespace py = pybind11;
 
+void addAXDIMMGlobalMethods(py::module& m) {
+  m.def("axdimm_starter", [](unsigned num_inference_engines) {
+    return static_cast<int>(axdimm::axdimm_starter(num_inference_engines));
+  });
+}
+
 PYBIND11_MODULE(caffe2_pybind11_state_axdimm, m) {
   m.doc() = "pybind11 stateful interface to Caffe2 workspaces - AXDIMM edition";
 
   addGlobalMethods(m);
+  addAXDIMMGlobalMethods(m);
   addObjectMethods(m);
   for (const auto& addition : PybindAdditionRegistry()->Keys()) {
     PybindAdditionRegistry()->Create(addition, m);
diff --git a/caffe2/python/workspace.py b/caffe2/python/workspace.py
index 079e6be1..72e543f9 100644
--- a/caffe2/python/workspace.py
+++ b/caffe2/python/workspace.py
@@ -96,6 +96,10 @@ if not has_gpu_support:
 
 if has_axdimm_support:
     AXDIMMDeviceType = caffe2_pb2.AXDIMM
+    AxdimmStarter = C.axdimm_starter
+
+if not has_axdimm_support:
+    AxdimmStarter = lambda: 0 # noqa
 
 IsNUMAEnabled = C.is_numa_enabled
 GetNumNUMANodes = C.get_num_numa_nodes
-- 
2.34.1

